Kernel/Bus/PCI/Device.cpp

```diff
@@ -61,4 +66,101 @@ void Device::disable_extended_message_signalled_interrupts()
     }
 }
 
+PCI::InterruptType Device::get_interrupt_type()
+{
+    return m_interrupts.m_type;
+}
+
+// Reserve `numbers_of_irqs` for this device. Returns the interrupt type
+// that was reserved. It is a noop for pin based interrupts as there
+// is nothing left to do. The second parameter `msi` is used by the
+// driver to indicate its intent to use message signalled interrupts.
+// MSI(x) is preferred over MSI if the device supports both.
+ErrorOr<InterruptType> Device::reserve_irqs(u8 number_of_irqs, bool msi)
+{
+    // Let us not allow partial allocation of IRQs for MSIx.
+    if (msi && is_msix_capable()) {
+        m_interrupts.m_start_irq = TRY(reserve_interrupt_handlers(number_of_irqs));
+        m_interrupts.m_irq_count = number_of_irqs;
+        m_interrupts.m_type = InterruptType::MSIX;
+        // If MSIx is available, disable the pin based interrupts
+        disable_pin_based_interrupts();
+        enable_extended_message_signalled_interrupts();
+    } else if (msi && is_msi_capable()) {
+        TODO();
+    }
+    return m_interrupts.m_type;
+}
+
+PhysicalAddress Device::msix_table_entry_address(u8 irq)
+{
+    auto index = irq - m_interrupts.m_start_irq;
+
+    VERIFY(index < m_interrupts.m_irq_count);
+    VERIFY(index >= 0);
+    auto table_bar_ptr = PCI::get_BAR(device_identifier(), static_cast<PCI::HeaderType0BaseRegister>(m_pci_identifier->get_msix_table_bar())) & PCI::bar_address_mask;
+    auto table_offset = m_pci_identifier->get_msix_table_offset();
+
+    return PhysicalAddress(table_bar_ptr + table_offset + (index * 16));
+}
+
+// This function is used to allocate an irq at an index and returns
+// the actual IRQ that was programmed at that index. This function is
+// mainly useful for MSI/MSIx based interrupt mechanism where the driver
+// needs to program. If the PCI device doesn't support MSIx interrupts, then
+// this function will just return the irq used for pin based interrupt.
+ErrorOr<u8> Device::allocate_irq(u8 index)
+{
+    if ((m_interrupts.m_type == InterruptType::MSIX) && is_msix_capable()) {
+        auto entry_ptr = TRY(Memory::map_typed_writable<msix_table_entry volatile>(msix_table_entry_address(index + m_interrupts.m_start_irq)));
+        entry_ptr->data = msi_data_register(m_interrupts.m_start_irq + index, false, false);
+        // TODO: we map all the IRQs to cpu 0 by default. We could attach
+        //  cpu affinity in the future where specific LAPIC id could be used.
+        u64 addr = msi_address_register(0, false, false);
+        entry_ptr->address_low = addr & 0xffffffff;
+        entry_ptr->address_high = addr >> 32;
+
+        u32 vector_ctrl = msix_vector_control_register(entry_ptr->vector_control, true);
+        entry_ptr->vector_control = vector_ctrl;
+        return m_interrupts.m_start_irq + index;
```

Comment:
This can overflow the `u8`

Reply:
Added an extra check plus changed the function to return `ErrorOr<u16>`

Reply:
Hmm, Yeah, you are right. I think this should be enough at the start of the program to avoid overflowing:
```
+    if ((m_interrupt_range.m_start_irq + index) >= GENERIC_INTERRUPT_HANDLERS_COUNT)
+        return Error::from_errno(EINVAL);
```